### DISCUSSION TITLE
Error when restoring a backup from one Docker container to another using pg_dump client

### DISCUSSION THREAD

**Discussion by igallart** [docker-library/postgres#1187] [General]
SO: Ubuntu 22.04 (32 GB RAM)

Install last client:
https://www.postgresql.org/download/linux/ubuntu/

Run a docker postgres:16

Backup from docker:
docker exec -t postgres pg_dump -U usr -Fc -Z4 -Od DATABASE > DATABASE.dump

Restore into docker with client:
pg_restore -vvv -U usr -h 127.0.0.1 -j 4 -Od TEST  DATABASE.dump

Show errors and fail:
pg_restore: allocating AH for TEST.dump, format 0
pg_restore: attempting to ascertain archive format
pg_restore: conectando a la base de datos para reestablecimiento
pg_restore: asumiendo reestablecimiento de sólo datos
pg_restore: entering restore_toc_entries_prefork
pg_restore: entering restore_toc_entries_parallel
memoria agotada

But the backup is done with pg_dump from client it works (pg_dump from localhost, not docker):
pg_dump -U usr -h 127.0.0.1 -Fc -Z4 -Od DATABASE > DATABASE.dump

**Comment by tianon** [comment]
How did you install Docker?  (`memoria agotada` is "memory exhausted", right?  that seems odd for a machine with 32GiB of RAM :thinking:)

**Reply by igallart** [reply]
Yes, is memory exhausted. Docker is installed how is say:
https://docs.docker.com/engine/install/ubuntu/#install-using-the-repository

**Comment by igallart** [comment]
Hello, I don't know if it's a mistake, but doing these steps one fails and another one doesn't.

`docker container run -d \
    --name postgres1 \
    --hostname postgres1 \
    --stop-timeout 20 \
    -e POSTGRES_DB=postgres \
    -e POSTGRES_PASSWORD=usr \
    -e POSTGRES_USER=usr \
    -e POSTGRES_PORT=5432 \
    -e POSTGRES_INITDB_ARGS="--locale=en_US.UTF-8 --lc-collate=C" \
    -p 5432:5432 \
    --restart unless-stopped \
    postgres:16
    
docker exec -it postgres1 psql -U usr -d postgres -c 'CREATE DATABASE "TEST";'
docker exec -it postgres1 psql -U usr -d TEST -c "CREATE TABLE users (id SERIAL PRIMARY KEY,name VARCHAR(50),age INT);"
docker exec -it postgres1 psql -U usr -d TEST -c "INSERT INTO users (name, age) VALUES ('Juan', 25), ('María', 30), ('Carlos', 22);"


docker container run -d \
    --name postgres2 \
    --hostname postgres1 \
    --stop-timeout 20 \
    -e POSTGRES_DB=postgres \
    -e POSTGRES_PASSWORD=usr \
    -e POSTGRES_USER=usr \
    -e POSTGRES_PORT=5432 \
    -e POSTGRES_INITDB_ARGS="--locale=en_US.UTF-8 --lc-collate=C" \
    -p 5433:5432 \
    --restart unless-stopped \
    postgres:16
    
docker exec -it postgres2 psql -U usr -d postgres -c 'CREATE DATABASE "TEST2";'


# Fail:
docker exec -t postgres1 pg_dump -U usr -Fc -Z4 -Od TEST > TEST.dump
pg_restore -vvv -p 5433 -U usr -h 127.0.0.1 -j 4 -Od TEST2 TEST.dump
pg_restore: allocating AH for TEST.dump, format 0
pg_restore: attempting to ascertain archive format
pg_restore: error: no se pudo leer desde el archivo de entrada: fin de archivo


# Work:
docker exec -it postgres1 bash
pg_dump -U usr -Fc -Z4 -Od TEST > TEST.dump
exit
docker cp postgres1:/root/TEST.dump .
pg_restore -vvv -p 5433 -U usr -h 127.0.0.1 -Od TEST2 TEST.dump


`

Maybe it's not the postgres client, but the ">" redirection. Is it possible?


libpq5:amd64                       16.1-1.pgdg22.04+1
pgadmin4-desktop               8.2               
pgadmin4-server                  8.2               
postgresql-client                   16+256.pgdg22.04+1
postgresql-client-16              16.1-1.pgdg22.04+1
postgresql-client-common    256.pgdg22.04+1

**Comment by igallart** [comment]
I think I found the issue or at least the solution, and it's changing -t to -i in docker exec:

docker exec -i postgres1 pg_dump -U usr -Fc -Z4 -Od TEST > TEST.dump

It now works! Thanks!

**Comment by LaurentGoderre** [comment]
Ah! Maybe the file was too big for a TTY

**Reply by igallart** [reply]
From what I could find out, with tty, the resulting file was created with copy instructions from the standard input, whereas with -i, it was only created as a standard dump file.

**Reply by tianon** [reply]
Ah yep, adding `-t` turns `\n` into `\r\n`:

```console
$ docker run --rm bash echo | hexdump -C
00000000  0a                                                |.|
00000001
$ docker run --rm -t bash echo | hexdump -C
00000000  0d 0a                                             |..|
00000002
```
